Libraries/LLM/Lora.swift

```diff
@@ -0,0 +1,545 @@
+            allLosses.append((losses * tokens).item(Float.self))
+            tokenCount += tokens.item(Int.self)
+
+            if batchCount != 0 && iteration + 1 >= batchCount {
+                break
+            }
+        }
+
+        return (sum(MLXArray(allLosses), stream: .cpu) / tokenCount).item(Float.self)
+    }
+
+    /// Given a ``LoRAModel`` with LoRA adaptors applied, load adapter weights from a `.safetensors` file.
+    public static func loadLoRAWeights(model: LoRAModel, url: URL) throws {
+        let weights = try ModuleParameters.unflattened(loadArrays(url: url))
+        try model.update(parameters: weights, verify: .noUnusedKeys)
+        eval(model)
+    }
+
+    /// Given a ``LoRAModel`` with LoRA adaptors applied, write adapter weights to a `.safetensors` file.
+    public static func saveLoRAWeights(model: LoRAModel, url: URL) throws {
+        let parameters = Dictionary(
+            uniqueKeysWithValues: model.trainableParameters().flattened())
+        try save(arrays: parameters, url: url)
+    }
+
+    public enum Progress: CustomStringConvertible {
+        case train(
+            iteration: Int, trainingLoss: Float, iterationsPerSecond: Double,
+            tokensPerSecond: Double)
+        case validation(iteration: Int, validationLoss: Float, validationTime: Double)
+        case save(iteration: Int, url: URL)
+
+        public var description: String {
+            switch self {
+            case .train(
+                let iteration, let trainingLoss, let iterationsPerSecond, let tokensPerSecond):
+                "Iteration \(iteration + 1): training loss \(trainingLoss.formatted()), "
+                    + "iterations/sec \(iterationsPerSecond.formatted()), "
+                    + "Tokens/sec \(tokensPerSecond.formatted())"
+            case .validation(let iteration, let validationLoss, let validationTime):
+                "Iteration \(iteration + 1): "
+                    + "validation loss \(validationLoss.formatted()), "
+                    + "validation time \(validationTime.formatted())s"
+            case .save(let iteration, let url):
+                "Iteration \(iteration + 1): saved weights to \(url.path())"
+            }
+        }
+    }
+
+    public enum ProgressDisposition {
+        case stop
+        case more
+    }
+
+    /// Train (or continue training) LoRA weights.
+    ///
+    /// - Parameters:
+    ///   - model: model to train
+    ///   - train: training dataset
+    ///   - validate: validate dataset
+    ///   - optimizer: optimizer used in training
+    ///   - loss: loss function
+    ///   - tokenizer: tokenizer
+    ///   - parameters: training parameters
+    ///   - progress: progress callback
+    public static func train(
+        model: LoRAModel, train: [String], validate: [String], optimizer: Optimizer,
+        loss: @escaping LoraLossFunction = loss, tokenizer: Tokenizer, parameters: Parameters,
+        progress: (Progress) async -> ProgressDisposition
```

Comment:
Rather than printing progress this is modeled as a callback that provides the information.  This might allow for some interesting applications:

- adaptive training that stops once a certain loss rate is hit
- integration in an example application, e.g. something running on iOS -- the training loop can be called by the UI code and received information about progress